File: LectureCodes/Filtering/periodictrends/Eigen/periodictrends.cpp
# include <cmath> // floor
# include <vector>
# include <string> // string, stod
# include <fstream> // getline
# include <Eigen/Dense>
# include <unsupported/Eigen/FFT>
# include <figure/figure.hpp>
using Eigen::VectorXd;
using Eigen::VectorXcd;

VectorXd read(const std::string file) {
  std::fstream fs(file); // filestream
  std::vector<double> data; // making use of push_back
  std::string buffer; // saving content of current line
  while (std::getline(fs, buffer)) { // 'for every line'
    data.push_back(std::stod(buffer)); // save number
  }
  // finally convert to eigen vector
  Eigen::Map<VectorXd> x(data.data(), data.size()); 
  return x;
}

int main() {
/* SAM_LISTING_BEGIN_0 */
  VectorXd x = read("trend.dat");
  const int n = x.size();
  
  mgl::Figure trend;
  trend.plot(x, "r");
  trend.title("Google: 'Vorlesungsverzeichnis'");
  trend.grid();
  trend.xlabel("week (1.1.2004-31.12.2010)");
  trend.ylabel("relative no. of searches");
  trend.save("searchdata");

  Eigen::FFT<double> fft;
  VectorXcd c = fft.fwd(x);
  VectorXd p = c.cwiseProduct(c.conjugate()).real()
                .segment(2, std::floor((n+1.)/2));

  // plot power spectrum
  mgl::Figure fig; fig.plot(p,"m"); fig.grid();
  fig.title("Fourier spectrum");
  fig.xlabel("index j of Fourier component");
  fig.ylabel("|c_j|^2");

  // mark 4 highest peaks with red star
  VectorXd p_sorted = p; 
  // sort descending
  std::sort(p_sorted.data(), p_sorted.data() + p.size(), std::greater<double>());
  for (int i = 0; i < 4; ++i) {
    // get position of p_sorted[i]
    int idx = std::find(p.data(), p.data()+p.size(), p_sorted[i]) - p.data();
    std::vector<double> ind = { double(idx + 1) }; // save in vectors
    std::vector<double> val = { p_sorted[i] }; // to be able to plot
    fig.plot(ind, val, " r*");
  }
  fig.save("fourierdata");
/* SAM_LISTING_END_0 */

  return 0;
}

File: LectureCodes/Filtering/periodictrends/Matlab/periodictrends.m
% Tracking of periodicity in data
% Data obtained from \textsf{Google Trends}, keyword ``Vorlesungsverzeichnis''
% Exported as .csv-file, non data lines removed, preprocessed by command \texttt{cut -f 2 -d,}

% read ASCI data from file
y = dlmread('trend.dat'); n = length(y);

figure('name','data');
plot(y,'r-'); grid on;
title('{\bf Google searches for keyword Vorlesungsverzeichnis}','fontsize',14);
xlabel('{\bf week (1.1.2004-31.12.2010)sa}','fontsize',14);
ylabel('{\bf relative no. of searches}','fontsize',14);
print -depsc2 '../PICTURES/searchdata.eps';

% Periodicity analysis by means of DFT
c = fft(y); 
p = abs(c(2:floor((n+1)/2))).^2; % Power spectrum
figure('name','Fourier spectrum');
plot(2:floor((n+1)/2),p,'m-'); grid on; hold on;
[mx,idx] = sort(p,'descend');
plot(1+idx(1:4),p(idx(1:4)),'rp');
xlabel('{\bf index j of Fourier component}','fontsize',14);
ylabel('{\bf |c_j|^2}','fontsize',14);
title('{\bf Energy spectrum of web search data}','fontsize',14);
print -depsc2 '../PICTURES/fourierdata.eps';
